File: src/main/java/project/rpg/ui/text/ScoreboardUI.java
package project.rpg.ui.text;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.DisplaySlot;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Score;
import org.bukkit.scoreboard.Scoreboard;
import project.rpg.textComponets.color.DefaultTextColors;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static net.kyori.adventure.text.Component.text;

public class ScoreboardUI {

    protected static final Map<Player, ScoreboardUI> _scoreboardMap = new HashMap<>();
    private final Scoreboard _scoreboard;
    private  Objective _objective;
    private final Player _player;

    public ScoreboardUI(Player player){
        _player = player;
        _scoreboard = Bukkit.getScoreboardManager().getNewScoreboard();

        _objective = _scoreboard.registerNewObjective("quest", "dummy", text("quest"));
        _objective.displayName(text("퀘스트 목록").color(DefaultTextColors.YELLOW.getColor()));

        _scoreboardMap.put(player, this);
    }

    public void setContents(List<String> contents){

        _objective.unregister();
        _objective = _scoreboard.registerNewObjective("quest", "dummy", text("quest"));
        _objective.displayName(text("퀘스트 목록").color(DefaultTextColors.YELLOW.getColor()));

        int priority = contents.size();
        Map<String, Integer> doubleCheck = new HashMap<>();
        for(String s : contents){
            priority--;

            if(doubleCheck.containsKey(s)) doubleCheck.put(s, doubleCheck.get(s)+1);
            else doubleCheck.put(s, 0);

            Score score = _objective.getScore(s + " ".repeat(Math.max(0, doubleCheck.get(s))));


            score.setScore(priority);
        }
        _player.setScoreboard(_scoreboard);
    }

    public void showScoreboard(){
        _objective.setDisplaySlot(DisplaySlot.SIDEBAR);
    }

    public void hideScoreboard(){
        _scoreboard.clearSlot(DisplaySlot.SIDEBAR);
    }

    public static ScoreboardUI getPlayer(Player player){
        return _scoreboardMap.get(player);
    }

}

File: src/main/kotlin/project/rpg/ui/text/ScoreboardUI.kt
package project.rpg.ui.text

import net.kyori.adventure.text.Component
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.scoreboard.DisplaySlot.SIDEBAR
import org.bukkit.scoreboard.Objective
import org.bukkit.scoreboard.Scoreboard
import project.rpg.textComponets.color.DefaultTextColors.YELLOW
import kotlin.math.max

class ScoreboardUI(private val _player: Player) {
    private val _scoreboard: Scoreboard
    private var _objective: Objective

    init {
        _scoreboard = Bukkit.getScoreboardManager().newScoreboard
        _objective = _scoreboard.registerNewObjective("quest", "dummy", Component.text("quest"))
        _objective.displayName(Component.text("퀘스트 목록").color(YELLOW.color))
        _scoreboardMap[_player] = this
    }

    fun setContents(contents: List<String>) {
        _objective.unregister()
        _objective = _scoreboard.registerNewObjective("quest", "dummy", Component.text("quest"))
        _objective.displayName(Component.text("퀘스트 목록").color(YELLOW.color))
        var priority = contents.size
        val doubleCheck: MutableMap<String, Int> = HashMap()
        for (s in contents) {
            priority--
            if (doubleCheck.containsKey(s)) doubleCheck[s] = doubleCheck[s]!! + 1 else doubleCheck[s] = 0
            val score = _objective.getScore(
                s + " ".repeat(
                    max(0.0, doubleCheck[s]!!.toDouble()).toInt()
                )
            )
            score.score = priority
        }
        _player.scoreboard = _scoreboard
    }

    fun showScoreboard() {
        _objective.displaySlot = SIDEBAR
    }

    fun hideScoreboard() {
        _scoreboard.clearSlot(SIDEBAR)
    }

    companion object {
        protected val _scoreboardMap: MutableMap<Player, ScoreboardUI> = HashMap()
        fun getPlayer(player: Player): ScoreboardUI? {
            return _scoreboardMap[player]
        }
    }
}
